File: PAT_Basic_Level_by_java/1059 C语言竞赛/1059 C语言竞赛.java
/**
 * 1059 C语言竞赛
 * 
 * 题目描述：
 * C语言竞赛的参赛者按排名获得不同奖品。给定参赛者人数 N 和他们的 ID，
 * 然后根据排名查询，输出对应的奖品：
 * - 排名第 1 的参赛者获得 "Mystery Award"（神秘大奖）
 * - 排名为素数的参赛者获得 "Minion"（小黄人）
 * - 其他参赛者获得 "Chocolate"（巧克力）
 * - 已被查询过的参赛者再次查询时输出 "Checked"
 * - 不存在的 ID 查询时输出 "Are you kidding?"
 * 
 * 实现原理：
 * 解题思路：
 * 1. 用 HashMap 存储 ID 到排名的映射关系，同时用数组 idArray 记录每个排名对应的 ID。
 * 2. 使用埃拉托色尼筛法（Sieve of Eratosthenes）预处理出 1~N 范围内的素数表。
 * 3. 对于每个查询，先检查 ID 是否存在，再检查是否已被查询过，最后根据排名输出对应奖品。
 * 
 * 算法核心逻辑：哈希表映射排名 → 素数筛预处理 → 按规则输出奖品。
 * 
 * 时间复杂度：O(N log log N + K)，空间复杂度：O(N)。
 */
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine().trim());

        // 存储 ID 到排名的映射，以及排名到 ID 的映射
        Map<String, Integer> rankMap = new HashMap<>();
        boolean[] checked = new boolean[N + 1]; // 标记某个排名是否被查询过
        String[] idArray = new String[N + 1];

        // 读取参赛者 ID，按顺序分配排名
        for (int i = 1; i <= N; i++) {
            String id = br.readLine().trim();
            rankMap.put(id, i);
            idArray[i] = id;
        }

        // 埃拉托色尼筛法，预处理 1~N 的素数表
        boolean[] isPrime = new boolean[N + 1];
        for (int i = 2; i <= N; i++) {
            isPrime[i] = true;
        }
        for (int i = 2; i * i <= N; i++) {
            if (isPrime[i]) {
                // 将 i 的倍数标记为非素数
                for (int j = i * i; j <= N; j += i) {
                    isPrime[j] = false;
                }
            }
        }

        // 处理查询
        int K = Integer.parseInt(br.readLine().trim());
        for (int i = 0; i < K; i++) {
            String id = br.readLine().trim();
            if (!rankMap.containsKey(id)) {
                // ID 不存在
                System.out.println(id + ": Are you kidding?");
            } else {
                int rank = rankMap.get(id);
                if (checked[rank]) {
                    // 已被查询过
                    System.out.println(id + ": Checked");
                } else {
                    checked[rank] = true;
                    if (rank == 1) {
                        System.out.println(id + ": Mystery Award"); // 第1名：神秘大奖
                    } else if (isPrime[rank]) {
                        System.out.println(id + ": Minion"); // 素数排名：小黄人
                    } else {
                        System.out.println(id + ": Chocolate"); // 其他：巧克力
                    }
                }
            }
        }
    }
}

File: PAT_Basic_Level_by_py3/1059 C语言竞赛/1059 C语言竞赛.py
# 题目：1059 C语言竞赛
#
# 题目描述：
#   C 语言竞赛是浙江大学计算机学院主持的一个欢乐的竞赛。既然竞赛主旨是为了好玩，颁奖规则也就制定得很滑稽：
#   - 0、冠军将赢得一份“神秘大奖”（比如很巨大的一本学生研究论文集……）。
#   - 1、排名为素数的学生将赢得最好的奖品 —— 小黄人玩偶！
#   - 2、其他人将得到巧克力。
#   给定比赛的最终排名以及一系列参赛者的 ID，你要给出这些参赛者应该获得的奖品。
#
# 输入格式：
#   输入第一行给出一个正整数 N（\le 10^4），是参赛者人数。随后 N 行给出最终排名，每行按排名顺序给出一位参赛者的 ID（4 位数字组成）。接下来给出一个正整数 K 以及 K 个需要查询的 ID。
#
# 输出格式：
#   对每个要查询的 ID，在一行中输出 `ID: 奖品`，其中奖品或者是 `Mystery Award`（神秘大奖）、或者是 `Minion`（小黄人）、或者是 `Chocolate`（巧克力）。如果所查 ID 根本不在排名里，打印 `Are you kidding?`（耍我呢？）。如果该 ID 已经查过了（即奖品已经领过了），打印 `ID: Checked`（不能多吃多占）。
#
# 实现原理：
#   - 使用集合(Set)进行去重和快速查找
#   - 循环迭代处理
#   - 列表操作
#
# 算法思路：
#   排名与查询：根据分数排名，
#   处理同分同名次的情况。
#
import sys

def is_prime(x):
    if x < 2:
        return False
    for divisor in range(2, int(x ** 0.5) + 1):
        if x % divisor == 0:
            return False
    return True


n = int(sys.stdin.readline())
rank = {sys.stdin.readline().strip(): i for i in range(1, n + 1)}
seen = set()
for _ in range(int(sys.stdin.readline())):
    x=sys.stdin.readline().strip()
    if x not in rank:
        print(f'{x}: Are you kidding?')
    elif x in seen:
        print(f'{x}: Checked')
    else:
        position = rank[x]
        award = (
            'Mystery Award' if position == 1
            else 'Minion' if is_prime(position)
            else 'Chocolate'
        )
        print(f'{x}: {award}')
        seen.add(x)
